### PR TITLE
【1.1.14】New User Configuration and Configuration Management for Linkis Management Console

### PR DESCRIPTION
**Pull request by v-kkhuang**
### What is the purpose of the change

linkis 管理台新增用户配置&配置管理功能（New user configuration and configuration management functions added to the Linkis management console）

### Related issues/PRs

Related User Configuration issues: #4841
Related Configuration Management issues: #4842

### Brief change log

- Define the core abstraction and interfaces of the EngineConn Factory;
- Define the core abstraction and interfaces of Executor Manager.

### Checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [Linkis mailing list](https://linkis.apache.org/community/how-to-subscribe) first)
- [ ] **If this is a code change**: I have written unit tests to fully verify the new behavior.

### CODE REVIEW

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java [hunk 1]
```diff
@@ -526,4 +532,197 @@ public Message deleteKeyValue(HttpServletRequest req, @RequestBody Map<String, O
...
+    long total = pageInfo.getTotal();
+    return Message.ok().data("configKey", list).data("totalPage", total);
+  }
+
+  @ApiOperation(value = "deleteBaseKeyValue", notes = "delete key", response = Message.class)
+  @ApiImplicitParams({@ApiImplicitParam(name = "id", required = true, dataType = "Integer")})
+  @RequestMapping(path = "/baseKeyValue", method = RequestMethod.DELETE)
+  public Message deleteBaseKeyValue(HttpServletRequest req, @RequestParam(value = "id") Integer id)
+      throws ConfigurationException {
+    checkAdmin(ModuleUserUtils.getOperationUser(req, "deleteBaseKeyValue"));
```
Comment (casionone) [inline linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:583]: 审计日志中 加上参数的id

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/dao/ConfigMapper.java [hunk 2]
```diff
@@ -71,4 +71,21 @@ List<ConfigKey> selectKeyByEngineTypeAndKeyList(
...
+  List<ConfigKey> getConfigEnBykey(
+      @Param("engineType") String engineType, @Param("key") String key);
+
+  List<ConfigUserValue> getUserConfigValue(
+      @Param("key") String key,
+      @Param("user") String user,
+      @Param("creator") String creator,
+      @Param("engineType") String engineType);
+
+  void insertKeyByBase(ConfigKey key);
```
Comment (casionone) [inline linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/dao/ConfigMapper.java:88]: 名字统一 ConfigKey configKey

Comment (v-kkhuang) [reply linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/dao/ConfigMapper.java:88]: done

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java [hunk 3]
```diff
@@ -526,4 +532,197 @@ public Message deleteKeyValue(HttpServletRequest req, @RequestBody Map<String, O
...
+    @ApiImplicitParam(name = "creator", required = false, dataType = "String", value = "creator"),
+    @ApiImplicitParam(name = "user", required = false, dataType = "String", value = "user"),
+    @ApiImplicitParam(name = "pageNow", required = false, dataType = "Integer", defaultValue = "1"),
+    @ApiImplicitParam(
+        name = "pageSize",
+        required = false,
+        dataType = "Integer",
+        defaultValue = "20"),
+  })
+  @RequestMapping(path = "/userkeyvalue", method = RequestMethod.GET)
```
[the remark this answers is not in the document]
Comment (v-kkhuang) [reply linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:694]: done

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java (hunk 1 above)
Comment (v-kkhuang) [reply linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:583]: done

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java [hunk 4]
```diff
@@ -526,4 +532,197 @@ public Message deleteKeyValue(HttpServletRequest req, @RequestBody Map<String, O
...
+    @ApiImplicitParam(
+        name = "enTreeName",
+        required = false,
+        dataType = "String",
+        value = "enTreeName")
+  })
+  @ApiOperationSupport(ignoreParameters = {"json"})
+  @RequestMapping(path = "/baseKeyValue", method = RequestMethod.POST)
+  public Message saveBaseKeyValue(HttpServletRequest req, @RequestBody Map<String, Object> json)
+      throws ConfigurationException, InstantiationException, IllegalAccessException {
```
[the remark this answers is not in the document]
Comment (v-kkhuang) [reply linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:639]: done

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java [hunk 5]
```diff
@@ -526,4 +528,195 @@ public Message deleteKeyValue(HttpServletRequest req, @RequestBody Map<String, O
...
+    PageHelper.startPage(pageNow, pageSize);
+    List<ConfigKey> list = null;
+    try {
+      list = configKeyService.getConfigBykey(engineType, key, req.getHeader("Content-Language"));
+    } finally {
+      PageHelper.clearPage();
+    }
+    PageInfo<ConfigKey> pageInfo = new PageInfo<>(list);
+    long total = pageInfo.getTotal();
+    return Message.ok().data("configKey", list).data("totalPage", total);
```
Comment (casionone) [inline linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:571]: 名字规范下 list or configKeyList 
觉得list 比较好

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java [hunk 6]
```diff
@@ -526,4 +528,195 @@ public Message deleteKeyValue(HttpServletRequest req, @RequestBody Map<String, O
...
+            .getOrCreateValidator(validateType)
+            .validate(defaultValue, validateRange)) {
+      String msg =
+          MessageFormat.format(
+              "Parameter configValue verification failed(参数defaultValue校验失败):"
+                  + "key:{0}, ValidateType:{1}, ValidateRange:{2},ConfigValue:{3}",
+              key, validateType, validateRange, defaultValue);
+      throw new ConfigurationException(msg);
+    }
+    Long id = Optional.ofNullable(configKey.getId()).orElse(0L);
```
Comment (casionone) [inline linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:661]: Long id = Optional.ofNullable(configKey.getId()).orElse(0L); 
这个获取是不是复杂了点 
直接获取后判空？

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java [hunk 7]
```diff
@@ -526,4 +528,195 @@ public Message deleteKeyValue(HttpServletRequest req, @RequestBody Map<String, O
...
+    PageHelper.startPage(pageNow, pageSize);
+    List<ConfigUserValue> list;
+    try {
+      list = configKeyService.getUserConfigValue(engineType, key, creator, user);
+    } finally {
+      PageHelper.clearPage();
+    }
+    PageInfo<ConfigUserValue> pageInfo = new PageInfo<>(list);
+    long total = pageInfo.getTotal();
+    return Message.ok().data("configValues", list).data("totalPage", total);
```
Comment (casionone) [inline linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:720]: 参数名 同样 如果不是多个list 列表 直接list 应该比较直接些

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java (hunk 5 above)
Comment (v-kkhuang) [reply linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:571]: done

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java (hunk 7 above)
Comment (v-kkhuang) [reply linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:720]: done

File: linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java (hunk 6 above)
Comment (v-kkhuang) [reply linkis-public-enhancements/linkis-configuration/src/main/java/org/apache/linkis/configuration/restful/api/ConfigurationRestfulApi.java:661]: done

**Pull request merged**